aten/src/ATen/native/cuda/ReduceSumProdKernel.cu

```diff
@@ -88,7 +88,7 @@ static void nansum_kernel_cuda(TensorIterator& iter) {
 
 static void prod_kernel_cuda(TensorIterator& iter) {
   auto general_dispatcher = [](TensorIterator& iter) {
-    AT_DISPATCH_ALL_TYPES(iter.dtype(), "prod_cuda", [&]() {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "prod_cuda", [&]() {
       prod_functor<scalar_t>{}(iter);
     });
```

Comment:
Add tests for complex prod.

Reply:
NOTE: backward for prod will not work for complex inputs, because it relies on `nonzero`, which is not yet implemented for the complex numbers.

Reply:
yes this sounds good let's update the backward formula in the next PR. This PR  https://github.com/pytorch/pytorch/pull/46596 enables backward for nonzero for complex. In general, to enable backward for an op for complex, you would need to add an entry in `GRADIENT_IMPLEMENTED_FOR_COMPLEX` in  `gen_variable.py`.